### PR TITLE
Bug 1795035: Give reason for not using cloud provider CA cert

### PR DESCRIPTION
**Pull request by mandre**
This makes it easier to debug in case CAPO can't read the CA cert from
the `openshift-config/cloud-provider-config` configmap.

### CODE REVIEW

File: pkg/cloud/openstack/clients/machineservice.go [hunk 1]
```diff
@@ -185,7 +185,8 @@ func NewInstanceServiceFromMachine(kubeClient kubernetes.Interface, machine *mac
 	}
 
 	cacert, err := getCACertFromConfigmap(kubeClient, "openshift-config", "cloud-provider-config", "ca-bundle.pem")
-	if err != nil || cacert == "" {
+	if err != nil {
```
Comment (pierreprinetti) [inline pkg/cloud/openstack/clients/machineservice.go]: watch out: the `cacert == ""` condition now leads to calling `NewInstanceServiceFromCloud` with a non-nil argument (`[]byte("")`, which is a non-nil slice of length zero). In `NewInstanceServiceFromCloud`, that will trigger the code for adding a new CA

https://github.com/openshift/cluster-api-provider-openstack/blob/86e93b0c54dbf72db95c7320548b9a22ef1193fa/pkg/cloud/openstack/clients/machineservice.go#L223-L237

**Comment by iamemilio** [comment]
We could check if the cloud object has a value for the `cafile` key, and we cant read the key from the configmap, then we can confidently throw an error if we have one without the other.

**Comment by iamemilio** [comment]
Perhaps we should validate the cert a little more here: https://github.com/openshift/cluster-api-provider-openstack/blob/393f7701ef3f5e7b6b86ba13221fb0b94d6214ec/pkg/cloud/openstack/clients/machineservice.go#L224

But that is separate from this issue/bug and can be its own BZ + PR

File: pkg/cloud/openstack/clients/machineservice.go (hunk 1 above)
Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go]: I think the way @mandre sets the warning is correct. Validation should be done here logically: https://github.com/openshift/cluster-api-provider-openstack/blob/393f7701ef3f5e7b6b86ba13221fb0b94d6214ec/pkg/cloud/openstack/clients/machineservice.go#L224

Comment (mandre) [reply pkg/cloud/openstack/clients/machineservice.go]: Yes, that was intentional. I think we shouldn't do validation on the cert and instead defer to the certPool.AppendCertsFromPEM(cert) call below to know what to do with an empty string.

It should be pretty obvious to debug if the user sets the configmap entry to an empty string by mistake.

Comment (iamemilio) [reply pkg/cloud/openstack/clients/machineservice.go]: Pierre made a good point in our conversation. The block of code wrapped in the conditional he linked above should only run if there is a cert. By default, we set the cert to "", so this means that this code block always runs by default, which is confusing.

Comment (adduarte) [reply pkg/cloud/openstack/clients/machineservice.go]: would it make sense to move the  cacert == "" to inside NewInstanceServiceFromCloud?

Comment (mandre) [reply pkg/cloud/openstack/clients/machineservice.go]: I'd rather not make any assumption about the value of cert other than it's not nil if certPool.AppendCertsFromPEM(cert) is capable of dealing with it.

**Pull request merged**